File: cses/02_sorting_searching/1629_movie-festival/solution.py
import sys


def main():
    # Đọc toàn bộ dữ liệu một lần rồi tách số (fast I/O).
    data = sys.stdin.buffer.read().split()
    if not data:
        return

    idx = 0
    n = int(data[idx]); idx += 1

    movies = []
    for _ in range(n):
        a = int(data[idx]); idx += 1
        b = int(data[idx]); idx += 1
        # Lưu (end, start) để sắp xếp tự nhiên theo thời điểm kết thúc.
        movies.append((b, a))

    # Sắp xếp tăng dần theo thời điểm kết thúc (greedy).
    movies.sort()

    count = 0
    cur_end = -1  # Thời điểm kết thúc của phim vừa chọn gần nhất.
    for end, start in movies:
        # Chọn phim nếu nó bắt đầu không sớm hơn lúc phim trước kết thúc.
        if start >= cur_end:
            count += 1
            cur_end = end

    sys.stdout.write(str(count) + "\n")


if __name__ == "__main__":
    main()

File: cses/02_sorting_searching/1629_movie-festival/solution.cpp
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    // Lưu mỗi phim dưới dạng (end, start) để sắp xếp tự nhiên theo thời điểm kết thúc.
    vector<pair<int, int>> movies;
    movies.reserve(n);
    for (int i = 0; i < n; ++i) {
        int start, end;
        cin >> start >> end;
        movies.emplace_back(end, start);
    }

    // Sắp xếp tăng dần theo thời điểm kết thúc (greedy).
    sort(movies.begin(), movies.end());

    int answer = 0;
    int current_end = -1;  // Thời điểm kết thúc của phim vừa chọn gần nhất.
    for (const auto& [end, start] : movies) {
        // Chọn phim nếu nó bắt đầu không sớm hơn lúc phim trước kết thúc.
        if (start >= current_end) {
            ++answer;
            current_end = end;
        }
    }

    cout << answer << '\n';
    return 0;
}
